crates/conversions/cairo-serde-macros/src/cairo_serialize.rs

```diff
@@ -0,0 +1,178 @@
+                            #prefix #prop
+                        }
+                    }
+                    Item::Enum => syn::parse_str::<syn::Ident>(&format!("field_{i}"))
+                        .unwrap()
+                        .to_token_stream(),
+                };
+
+                quote_spanned! {f.span()=>
+                    conversions::serde::serialize::CairoSerialize::serialize(& #name, writer);
+                }
+            });
+
+            quote! {
+                #(#recurse),*
+            }
+        }
+        Fields::Unit => TokenStream::new(),
+    }
+}
+
+fn destruct_fields(fields: &Fields) -> TokenStream {
+    match fields {
+        Fields::Named(fields) => {
+            let recurse = fields.named.iter().map(|f| {
+                let name = &f.ident;
+
+                quote_spanned! {f.span() =>
+                    #name
+                }
+            });
+
+            quote! {
+                {#(#recurse),*}
+            }
+        }
+        Fields::Unnamed(fields) => {
+            let recurse = fields.unnamed.iter().enumerate().map(|(i, f)| {
+                let name = syn::parse_str::<syn::Ident>(&format!("field_{i}")).unwrap();
+
+                quote_spanned! {f.span() =>
+                    #name
+                }
+            });
+
+            quote! {
+                (#(#recurse),*)
+            }
+        }
+        Fields::Unit => TokenStream::new(),
+    }
+}
+
+// creates code for `CairoSerialize::serialize` body
+fn create_func_body(data: &Data, span: &TokenStream) -> TokenStream {
+    match data {
+        Data::Struct(data) => {
+            call_trait_on_field(&data.fields, Item::Struct)
+        },
+        Data::Enum(data) => {
+            // generate match arms by matching on next integer literals (discriminator)
+            // then generate trait calls for variants fields
+            let arms = data.variants.iter().enumerate().map(|(i,variant)| {
+                let name = &variant.ident;
+                let calls = call_trait_on_field(&variant.fields, Item::Enum);
+                let destructurization = destruct_fields(&variant.fields);
+                let lit = syn::parse_str::<syn::LitInt>(&format!("{i}_u32")).unwrap();
+
+                quote! {
+                    Self::#name #destructurization => {
+                        conversions::serde::serialize::CairoSerialize::serialize(&#lit, writer);
+                        #calls
+                    }
+                }
+            });
+
+            // empty match does not work with references
+            // and `self` is behind reference
+            if data.variants.is_empty() {
+                quote! {}
+            }else{
+                quote! {
+                    match self {
+                        #(#arms,)*
+                    };
+                }
+            }
+
+        }
+        // can not determine which variant should be used
+        // use enum instead
```

Comment:
Union can not be serialized/deserialized because there is no way to determine which variant should be used. Enum is prefered as it contains discriminator